### PR TITLE
Timing out packets on ordered channels when in flushing state aborts upgrade and closes channel

### PR DESCRIPTION
**Pull request by chatton**
## Description

closes: #4454

This PR updates TimeoutExecuted to ensure that when an ordered channel is passed

1. The channel is closed
2. The upgrade info is deleted
3. An error receipt is written

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [x] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [x] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

File: modules/core/04-channel/keeper/timeout_test.go [hunk 1]
```diff
@@ -444,6 +444,51 @@ func (suite *KeeperTestSuite) TestTimeoutExecuted() {
...
+					Timeout: types.NewTimeout(clienttypes.ZeroHeight(), 1),
+				})
+			},
+			func(packetCommitment []byte, err error) {
+				suite.Require().NoError(err)
+
+				channel := path.EndpointA.GetChannel()
+				suite.Require().Equal(types.CLOSED, channel.State, "channel state should be CLOSED")
+
+				upgrade, found := suite.chainA.App.GetIBCKeeper().ChannelKeeper.GetUpgrade(suite.chainA.GetContext(), path.EndpointA.ChannelConfig.PortID, path.EndpointA.ChannelID)
```
Comment (DimitrisJim) [inline modules/core/04-channel/keeper/timeout_test.go:479]: possible future refactor: checking post-state for cleared upgrade state is done commonly iirc, could be nice if possible to have a helper for this.

File: modules/core/04-channel/keeper/timeout.go [hunk 2]
```diff
@@ -171,17 +164,26 @@ func (k Keeper) TimeoutExecuted(
...
+			} else if !k.HasInflightPackets(ctx, packet.GetSourcePort(), packet.GetSourceChannel()) {
+				// set the channel state to flush complete if all packets have been flushed.
 				channel.State = types.STATE_FLUSHCOMPLETE
 				k.SetChannel(ctx, packet.GetSourcePort(), packet.GetSourceChannel(), channel)
 			}
 		}
 	}
 
+	if channel.Ordering == types.ORDERED {
+		// note: we continue to close the unordered channel even if the upgrade has been aborted.
```
Comment (DimitrisJim) [inline modules/core/04-channel/keeper/timeout.go:176]:
```suggestion
		// note: we continue to close the ordered channel even if the upgrade has been aborted.
```

File: modules/core/04-channel/keeper/timeout.go [hunk 3]
```diff
@@ -150,13 +150,6 @@ func (k Keeper) TimeoutExecuted(
...
-	if channel.Ordering == types.ORDERED {
-		channel.State = types.CLOSED
-		k.SetChannel(ctx, packet.GetSourcePort(), packet.GetSourceChannel(), channel)
-		emitChannelClosedEvent(ctx, packet, channel)
-	}
-
-	// TODO: handle situation outlined in https://github.com/cosmos/ibc-go/issues/4454
 	// if an upgrade is in progress, handling packet flushing and update channel state appropriately
 	if channel.State == types.STATE_FLUSHING {
 		counterpartyUpgrade, found := k.GetCounterpartyUpgrade(ctx, packet.GetSourcePort(), packet.GetSourceChannel())
```
Comment (damiannolan) [inline modules/core/04-channel/keeper/timeout.go:155]: Let's say the timeout on `counterpartyUpgrade` is not set yet (TRY has been called, we are in state `FLUSHING` but counterparty timeout is not yet set), but we are in the process of timing out a packet on an ordered channel. 

We would end up skipping the code in `timeout.IsValid()`, timing out the packet and closing the channel, but the upgrade state would still remain in store.

Comment (DimitrisJim) [reply modules/core/04-channel/keeper/timeout.go:155]: guess we should be able to conditionally delete in the `if channel.Ordering == types.ORDERED` block yea?

Comment (chatton) [reply modules/core/04-channel/keeper/timeout.go:155]: really nice catch, I guess then the situation we can be in is that

timeout is invalid, LastSequenceSend is 0, but upgradeFields are set. We could use this as the check for also triggering the abort.

**Review by damiannolan — APPROVED** [review]
Discussed together synchronously and came to a solution that works. Let's revisit this and see if it can be cleaned up or improved during the audit. 

https://github.com/cosmos/ibc-go/pull/4475/commits/680ce92b9858a4cd1d8ad4f3d6d97315ccc534e3 contains the changes to allow aborting an upgrade on an ordered channel when timing out a packet and ending up in a closed state.

**Pull request merged**